fineract-investor/src/main/java/org/apache/fineract/investor/service/ExternalAssetOwnerLoanProductAttributesWriteServiceImpl.java

```diff
@@ -0,0 +1,187 @@
+                Arrays.asList(ExternalAssetOwnerLoanProductAttributeRequestParameters.ATTRIBUTE_KEY,
+                        ExternalAssetOwnerLoanProductAttributeRequestParameters.ATTRIBUTE_VALUE));
+        final Type typeOfMap = new TypeToken<Map<String, Object>>() {}.getType();
+        fromApiJsonHelper.checkForUnsupportedParameters(typeOfMap, apiRequestBodyAsJson, requestParameters);
+
+        final List<ApiParameterError> dataValidationErrors = new ArrayList<>();
+        final DataValidatorBuilder baseDataValidator = new DataValidatorBuilder(dataValidationErrors).resource("loanproductattribute");
+
+        baseDataValidator.reset().parameter(ExternalAssetOwnerLoanProductAttributeRequestParameters.ATTRIBUTE_KEY).value(attributeKey)
+                .notBlank().notExceedingLengthOf(255);
+
+        baseDataValidator.reset().parameter(ExternalAssetOwnerLoanProductAttributeRequestParameters.ATTRIBUTE_VALUE).value(attributeValue)
+                .notBlank().notExceedingLengthOf(255);
+
+        if (!dataValidationErrors.isEmpty()) {
+            throw new PlatformApiDataValidationException("validation.msg.validation.errors.exist", "Validation errors exist.",
+                    dataValidationErrors);
+        }
+    }
+
+    private void validateLoanProductExistsAndAttributeDoesNotExist(Long loanProductId, String attributeKey) {
+        validateLoanProductExists(loanProductId);
+        validateLoanProductAttributeDoesNotExist(loanProductId, attributeKey);
+    }
+
+    private void validateLoanProductExists(Long loanProductId) {
+        if (!loanProductRepository.existsById(loanProductId)) {
+            throw new LoanProductNotFoundException(loanProductId);
+        }
+    }
+
+    private void validateLoanProductAttributeDoesNotExist(Long loanProductId, String attributeKey) {
+        if (externalAssetOwnerLoanProductAttributesRepository.existsByLoanProductIdAndKey(loanProductId, attributeKey)) {
+            throw new ExternalAssetOwnerLoanProductAttributeAlreadyExistsException(
+                    "attributeKey already exists for the loanProductId: " + loanProductId + ". Use PUT call to UPDATE the attribute.");
+        }
+    }
+
+    private void validateLoanProductAttributeKeysMatch(String attributeKeyFromRequest, String attributeKeyFromDB) {
+        if (!attributeKeyFromRequest.equals(attributeKeyFromDB)) {
+            throw new ExternalAssetOwnerLoanProductAttributesException(
+                    "The attribute key of requested update attribute does not match the attribute key from database.");
+        }
+    }
+
+    private void validateExternalAssetOwnerLoanProductAttribute(String attributeKey, String attributeValue) {
+        for (Class<?> implementingClass : implementingClasses) {
+            if (implementingClass.isEnum()) {
+                for (Object obj : implementingClass.getEnumConstants()) {
+                    ExternalAssetOwnerLoanProductAttribute objEnum = (ExternalAssetOwnerLoanProductAttribute) obj;
```

Comment:
Isnt this a little weird? Via reflection fetching all the enums and trying to resolve attribute keys? Also its a set, the order is not guaranteed. So in case there are overlapping enums it is not deterministic which will be used. Also I am not sure whether we really need such flexibility and "dynamic resolution" here...

Reply:
This way 3rd parties can add new enum values via a custom module. Otherwise it won't be possible and we can't do any validation on them (which in turn hurts data integrity).